### PR TITLE
Chore/example back on list item

### PR DESCRIPTION
**Pull request by alxlr**
- Return to first item of virtualized list on back button press

https://www.loom.com/share/80099727b07e43e2ac1f088be8f47b2a?sid=a70d798c-29c2-44bb-aa0e-45c271c7e3b4

### CODE REVIEW

File: packages/example/src/components/remote-control/OverridableEventEmitter.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    for (let index = this.handlers[eventName].length - 1; index >= 0; index--) {
+      const handler = this.handlers[eventName][index];
+      if (handler(event)) {
+        return;
+      }
+    }
+  };
+}
+
+export default new OverridableEventEmitter();
```
Comment (pierpo) [inline packages/example/src/components/remote-control/OverridableEventEmitter.ts:25]: you're exposing a singleton : you should only expose the class, and instantiate it in the RemoteControlManager ;) (like we used to do with mitt)

File: packages/example/src/components/remote-control/OverridableEventEmitter.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { SupportedKeys } from './SupportedKeys';
+
+class OverridableEventEmitter {
+  private handlers = new Map<string, ((event: SupportedKeys) => boolean)[]>();
```
Comment (pierpo) [inline packages/example/src/components/remote-control/OverridableEventEmitter.ts:4]: Ideally, we'd like to type like mitt does right now. Mitt allows to type what are the events and for each event what is the param signature. You should have a look at how mitt does it :)

```suggestion
  private handlers = new Map<T, U[]>();
```

File: packages/example/src/modules/program/view/ProgramList.tsx [hunk 3]
```diff
@@ -45,46 +50,67 @@ export const ProgramList = ({
...
+      ) {
+        return false;
+      }
+      listRef.current.focus(0);
+      return true;
+    };
+    RemoteControlManager.addKeydownListener(remoteControlListener);
+
+    return () => RemoteControlManager.removeKeydownListener(remoteControlListener);
+  }, [isActive, listRef, navigation]);
```
Comment (pierpo) [inline packages/example/src/modules/program/view/ProgramList.tsx:68]: Maybe the time has come to create a hook `useKey(SupportedKeys.Back, () => { yourWholeLogic() })` 😁

File: packages/example/src/components/remote-control/OverridableEventEmitter.ts (hunk 2 above)
Comment (alxlr) [reply packages/example/src/components/remote-control/OverridableEventEmitter.ts:4]: Yeah, I saw how they handled it in mitt but I figured we didn't need the whole implementation for our use case :-) Can we take a look at it together quickly so we can agree on what to and what not to implement?

File: packages/example/src/modules/program/view/ProgramList.tsx (hunk 3 above)
Comment (alxlr) [reply packages/example/src/modules/program/view/ProgramList.tsx]: Can I use a custom hook like this in a useEffect? :o Interested to discuss this

File: packages/example/src/components/remote-control/OverridableEventEmitter.ts (hunk 2 above)
Comment (alxlr) [reply packages/example/src/components/remote-control/OverridableEventEmitter.ts]: Done 👍🏽

File: packages/example/src/modules/program/view/ProgramList.tsx (hunk 3 above)
Comment (alxlr) [reply packages/example/src/modules/program/view/ProgramList.tsx]: Done :-)

File: packages/example/src/components/remote-control/OverridableEventEmitter.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
+export type EventType = string | symbol;
```
Comment (pierpo) [inline packages/example/src/components/remote-control/OverridableEventEmitter.ts:1]: Add a comment here that you took inspiration from mitt

File: packages/example/src/components/remote-control/OverridableEventEmitter.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+// An array of all currently registered event handlers for a type
+export type EventHandlerList<T = unknown> = Array<Handler<T>>;
+
+// A map of event types and their corresponding event handlers.
+export type EventHandlerMap<Events extends Record<EventType, unknown>> = Map<
+  keyof Events,
+  EventHandlerList<Events[keyof Events]>
+>;
+
+export default class OverridableEventEmitter<Events extends Record<EventType, unknown>> {
```
Comment (pierpo) [inline packages/example/src/components/remote-control/OverridableEventEmitter.ts:16]:
```suggestion
/** 
 * This event emitter is a minimal reimplementation of `mitt` with the support of stoppable event propagation
 */
export default class CustomEventEmitter<Events extends Record<EventType, unknown>> {
```

File: packages/example/src/hooks/useKey.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { useEffect } from 'react';
+import RemoteControlManager from '../components/remote-control/RemoteControlManager';
+import { SupportedKeys } from '../components/remote-control/SupportedKeys';
+
+export const useKey = (key: SupportedKeys, callback: (pressedKey: SupportedKeys) => boolean) => {
```
Comment (pierpo) [inline packages/example/src/hooks/useKey.ts:5]: Maybe we can add a little tsdoc ?

```suggestion
/**
 * A convenient hook to listen to a key and react to it
 *
 * @example useKey(SupportedKeys.Back, () => { console.log('pressed back!') })
 */
export const useKey = (key: SupportedKeys, callback: (pressedKey: SupportedKeys) => boolean) => {
```

**Review by pierpo — APPROVED** [review]
I took over the missing pieces, and solved the blocking bug in another PR.
Great job, thanks! 🎉

**Pull request merged**